### PR TITLE
Large Refactor, adding sound

### PR DESCRIPTION
**Pull request by Joeyh021**
I've done a few bits:
- Split it into two crates, the implementation and the library. These will need to be published separetely on crates.io
- Reworked the public API to use a `Pixel` enum instead of `u8`s, becuase type systems 
  - This is a breaking change, and I'll bump the major version when I release
- Added audio, based on the cpal buzzer example
  - This can be tested with the test rom with sound, it should emit a 1 second beep
- Added a small warning to be emitted when the interpreter is running slow

Still to do:
- add better error handling
- add thorough logging, to log frametimes, errors, etc
- add some sort of benchmark
- write some tests. This is more of a want than a must, but it would be helpful to detect regressions.

### CODE REVIEW

File: chip8-base/src/interpreter/display.rs [hunk 1]
```diff
@@ -29,14 +29,13 @@ pub fn init() -> Option<(EventLoop<()>, Window, Pixels)> {
...
+pub fn update(pixels: &mut Pixels, buffer: &[[crate::Pixel; 64]; 32]) {
     let mut old_buf = pixels.get_frame();
     for px in buffer.concat() {
         old_buf
             .write_all(match px {
-                0 => &[0_u8, 0_u8, 0_u8, 255_u8],
-                1 => &[255_u8, 255_u8, 255_u8, 255_u8],
-                _ => unreachable!(),
+                crate::Pixel::Black => &[0_u8, 0_u8, 0_u8, 255_u8],
+                crate::Pixel::White => &[255_u8, 255_u8, 255_u8, 255_u8],
```
Comment (ericthelemur) [inline chip8-base/src/interpreter/display.rs:38]: Maybe move this match into a From<[u8; 4]> for Pixel. Might not work with the borrow though

File: chip8-base/src/interpreter/sound.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            }
+        }
+    };
+
+    let stream = device.build_output_stream(config, data_fn, err_fn)?;
+
+    stream.play()?;
+
+    Ok(stream)
+}
```
Comment (ericthelemur) [inline chip8-base/src/interpreter/sound.rs]: Imagine wholesale ripping of branches/repos, couldn't be me ....

File: chip8-base/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+/// CHIP-8 interpreters can be built using this trait.
+/// [`step`][Interpreter::step] should be implemented on a type representing a CHIP-8 Interpreter to run the interpreter one clock cycle at a time, such that calling it in a loop runs the interpreter.
+pub trait Interpreter {
+    /// Executes the next CHIP-8 Instruction, modifying the state of the virtual machine/CPU accordingly.
+    /// This is the main driver function, running the interpreter one clock cycle at a time.
+    /// # Return
+    /// If the instruction modified the state of the display, then an updated [`Display`][Display] should be returned.
+    /// # Panics
+    /// Should panic if an unrecognised instruction is encountered
```
Comment (ericthelemur) [inline chip8-base/src/lib.rs:26]: Definitely feel like this bit of the spec will be forgotten (and mismatches current)

**Review by ericthelemur — APPROVED** [review]

File: chip8-base/src/interpreter/sound.rs (hunk 2 above)
Comment (Joeyh021) [reply chip8-base/src/interpreter/sound.rs]: I should really double check the licensing actually

File: chip8-base/src/lib.rs (hunk 3 above)
Comment (Joeyh021) [reply chip8-base/src/lib.rs:26]: doesnt mismatch, current one panics at the decode stage

File: chip8-base/src/interpreter/display.rs (hunk 1 above)
Comment (Joeyh021) [reply chip8-base/src/interpreter/display.rs:38]: Hmmm, I don't think it's necessary to abstract out, it's only used here.

**Pull request merged**